### ISSUE TITLE
Predicate for file extensions

### ISSUE THREAD

**Issue by SwiftsNamesake** [haskell/filepath#60]
Are there any plans to add a predicate for specific file extensions (like `hasExtension` but taking an additional argument)? Any reason why it's been omitted? Would you accept a pull request for it?

The only hurdle is coming up with a good name (since `hasExtension` is already taken).

**Comment by hvr** [comment]
Did somebody say bikeshedding? :)

We already have binary operators `<.>` and `-<.>`, what about

```hs
(?<.>) :: FilePath -> String -> Bool
```

or maybe more in tradition with lisp style predicate naming:

```hs
(<.>?) :: FilePath -> String -> Bool
```


? =)

**Comment by ndmitchell** [comment]
Any reason to go for an operator? I'd have gone with isExtension, but there the issue of argument order - isSuffixOf suggests the extension should go first.

**Comment by hvr** [comment]
No particular reason; it just seemed like it would have completed the existing family of `FilePath -> String -> x` binary operators...

**Comment by SwiftsNamesake** [comment]
I like to delegate my bikeshedding to a select committee :P
I'd have submitted a pull request right away, but it's such a glaring omission that I wondered if there was some weird reason why it's not in the library already...

**Comment by SwiftsNamesake** [comment]
I suppose `hasTheExtensions` would work. I think @ndmitchell's suggestion is too generic.

**Comment by ndmitchell** [comment]
Joining words like `the` and `and` in function names is always a bit nasty (looking at `zipWithAndUnzipM` in particular). I still quite like `isExtension`, since it matches `isSuffixOf` and `isPrefixOf`, which is very much a similar function.

The reason this function hasn't been added is because `takeExtension x == ".foo"` is usually fine. However, that does require you to know you need to put `.` in front of the `foo`, which is why I think a helper that implicitly does that as required is simpler.

**Comment by hvr** [comment]
Fwiwl, if you want to use `filter`, then your options currently are more or less

    filter (\fn -> takeExtension fn == ".foo")
    filter ((== ".foo") . takeExtension)

However, with the predicate (with an argument order like `isSuffixOf`)

    isExtensionOf :: String -> FilePath -> Bool

this use-case would IMO look nicer, i.e.

    filter (isExtensionOf "foo")

With a `<.>? :: FilePath -> String ->Bool` operator, it'd b

    filter (<.>? "foo")

which isn't that bad either imho (but I'm biased, because I'm a fan of binary operators)  :-)

**Comment by ndmitchell** [comment]
BTW, I haven't seen a pull request, so not sure where it has gone. That said, I'd figure out a name first, since the actual pull request should be quite simple.

**Comment by ndmitchell** [comment]
I tend to reserve binary operators for things that are super common, as people pretty much have to "learn" them - they can't guess from the name. I can't imagine this will be super common. Agreed about the filter case - so I guess there are two reasons to include this function - both good reasons - so adding something makes sense.

Thinking further, I like `isExtensionOf` more and more.

**Comment by hvr** [comment]
`isExtensionOf` is clearly the less controversial choice here :-)

**Comment by SwiftsNamesake** [comment]
Yeah, I agree with @ndmitchell that it looks weird. I'd like to weigh in in favour of `isExtensionOf`, which makes sense when used as an operator.

**Comment by SwiftsNamesake** [comment]
I think `isSuffixOf` is too generic. Shouldn't we use a name that is specific to file extensions? There's already a commonly used function with that name. Also, it only suggests the extension should go first if you imagine it being used as inline (eg. `a \`isSuffixOf\` b`)

**Comment by ndmitchell** [comment]
Great, seems like we're all in agreement on `isExtensionOf :: String -> FilePath -> Bool`. Pull request welcome!

**Comment by SwiftsNamesake** [comment]
@ndmitchell It's done. Take a look!

**Comment by ndmitchell** [comment]
So the pull request is at #61. @SwiftsNamesake suggested we take the discussion of semantics back here (I'm not fussed whether we discuss these things on the pull request of the bug report, but this issue is fine).

We're all agreed on:

    isExtensionOf :: String -> FilePath -> Bool
    isExtensionOf "jpg" "foo.jpg" = True

What we're not agreed on is:

    isExtensionOf ".gz" "bar/foo.tar.gz"
    isExtensionOf "tar.gz" "bar/foo.tar.gz"
    isExtensionOf ".tar.gz" "bar/foo.tar.gz"
    
I suggest they should all work. Adding an initial leading dot or not is exactly what people have to overthink in the current situation. No one who passes `isExtensionOf ".gz"` expects that to be `const False`.

Secondly, should it go up multiple extensions. Again - principle of least surprise - if you ask someone "is that the extension" they'll say yes. I agree this is the most flexible interpretation possible, but that seems right in this instance.

**Comment by SwiftsNamesake** [comment]
@ndmitchell I actually don't have a strong opinion on this matter, except for consistency (cf. `splitExtension`, which gives you the *last extension* (including the separator).

**Comment by hvr** [comment]
Fwiw, I tend to agree w/ @ndmitchell, what I'm not sure is what

```hs
isExtensionOf ".gz" "bar/foo.tar..gz"
isExtensionOf "..gz" "bar/foo.gz"
isExtensionOf "..gz" "bar/foo.tar.gz"
isExtensionOf "tar.gz" "bar/foo.tar..gz"
isExtensionOf ".tar.gz" "bar/foo.tar..gz"
```

is supposed to return.

**Comment by SwiftsNamesake** [comment]
@hvr I don't think we should be *too* forgiving either. I think @ndmitchell's suggestion is perfectly reasonable, as well as being easy to implement.

```haskell
isExtensionOf ext = isPrefixOf ext . snd . break (== '.') -- Alternatively, dropWhile (/= '.')
```
The only drawback is that this doesn't reject malformed paths (not sure how this library tends to deal with that).

**Comment by ndmitchell** [comment]
I agree, don't be too forgiving. In particular, I think `isExtensionOf "bar/foo" "bar/foo"` should be False - you've clearly gone way past an extension. I think:

    isExtensionOf ext file = dots ext `isSuffixOf` dots (takeExtensions file)
        where dots ('.':xs) = splitOn "." xs
              dots xs = splitOn "." xs

Probably captures all the semantics we want, assuming `splitOn` from the `extra` library (which should definitely NOT be added as a dependency).

**Comment by SwiftsNamesake** [comment]
Maybe this is horribly pedantic of me (even for a programmer), but since you brought up the principle of least surprise: given the semantics you describe, wouldn't `areExtensionsOf` make more sense? We already have `takeExtension` and `takeExtensions`, with analogous behaviour.

Wouldn't breaking that pattern violate that principle?

------

Regarding @ndmitchell's implementation, `breakOn` allows adjacent separators, which is not what we want, is it? How does `filepath` deal with malformed input, in general? I'd say it makes sense for `isExtensionOf` to rely on the behaviour of `takeExtensions`.

I suggest the following refinement of the code I posted earlier:

```haskell
isExtensionOf ext = isSuffixOf ext . takeExtensions
```

**Comment by ndmitchell** [comment]
Note that `addExtension` allows multiple extensions, so it's not unprecedented. I also don't like the `are` to agree with extensions since it's inconsistent with `isPrefixOf`. You could argue for `isExtensionsOf`, but it gets very messy and ungrammatical. I can see why `isExtensionOf` isn't quite consistent, but I can't think of anything better.

My code used `splitOn`, which treats multiple separators properly, but your code is way simpler. Your code has the issue that `isExtensionOf "o" "test.foo"` returns True. I think you can solve that by optionally adding `.` to `ext` if it isn't already there, and then I think it works properly.

**Comment by SwiftsNamesake** [comment]
@ndmitchell `splitOn` will not reject eg. "tar..g". Hence my point about rejecting invalid paths. It *is* invalid right? Or does Unix allow that sort of thing?

```haskell
> splitOn "." "foo.tar..gz"
["foo","tar","","gz"]
```

Good catch regarding partial matches, I'll work something out.

**Comment by ndmitchell** [comment]
I think Linux allows it - I wouldn't overly worry though - although `isExtensionOf ".tar..gz" "foo.tar..gz" == True` seems correct.

**Comment by SwiftsNamesake** [comment]
@ndmitchell Regarding the name, I think it's fine as long as we document the behaviour properly (which we're obviously going to do).

@hvr I'll submit another pull request and await your response. I suggest the following implementation (which has with the semantics that @ndmitchell describes):

```haskell
-- | Looks good?
isExtensionOf :: String -> FilePath ->FilePath
isExtensionOf ext@('.':_) = isSuffixOf ext . takeExtensions
isExtensionOf ext         = isSuffixOf ('.':ext) . takeExtensions
```

**Comment by ndmitchell** [comment]
Looks good to me.

**Comment by ndmitchell** [comment]
Fixed as #61.
